// manifest.json/Gruntfile.js
/*
 * manifest.json
 * https://github.com/michaelp983/manifest.json
 *
 * Copyright (c) 2013 Michael Brady Paulson
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-copy');

    // Project configuration.
    grunt.initConfig({
        // Configuration to be run (and then tested).
        manifest_json: {
            all: {
                options: {
                    dist: './dist/',
                    all: true
                }
            }
        },
        copy: {
            all: {
                files: [
                    {src: ['./demo.html'], dest: './app-debug/demo.html'},
                    {expand: true, src: ['./dist/**'], dest: 'app-debug/'}
                ]
            }
        },

        watch: {
            all: {
                files: [
                    './core/**/*.js',
                    './core/js-bsdc/**/*.js',
                    './demo.html'
                ],
                tasks: ['manifest_json']
            }
        }
    });

    // Actually load this plugin's task(s).
    grunt.loadTasks('tasks');

    // By default, lint and run all tests.
    grunt.registerTask('default', ['manifest_json', 'watch']);
};
